### PR TITLE
Copter: do not enter airmode when armed with switch

### LINKED ISSUE #17836: Copter: arming via switch enables airmode - but no way to disable
## Bug report

**Issue details**

Arming Copter via switch (RCx_OPTION = 41) also activates AirMode in Stabilize flight mode. This is documented behaviour [as per the wiki](https://ardupilot.org/copter/docs/airmode.html).

However there is no way to **disable/prevent** AirMode from activating in this scenario. You cannot use Stabilize without AirMode when arming via switch.

**Version**
4.0.7

**Platform**
[  ] All
[  ] AntennaTracker
[X] Copter
[  ] Plane
[  ] Rover
[  ] Submarine

**Airframe type**
Tested on Y6B, but likely applies to all

**Hardware type**
Orange Cube

**Logs**

### PR DESCRIPTION
**Pull request by IamPete1**
Fixes #17836

This stops entering airmode with armed with a switch. If you want airmode in ACRO you can set `ACRO_OPTIONS`. Or configure the airmode on/off switch. Or use a interlock switch. 

The current code makes some less keen to use switch arming due to being stuck with air mode. 

We do loose the ability to auto enable in stabilize mode, instead of removing we could put this under `FLIGHT_OPTIONS`. We could add a new arming switch `arm/disarm with airmode`, that might be slightly nicer than adding a `FLIGHT_OPTIONS` bit. 

Quite hard to change this without change behavior for some.

### CODE REVIEW

**Review by andyp1per — APPROVED** [review]

**Comment by tridge** [comment]
we discussed on the EU dev call, and had some Qs:
 - why would you not want airmode?
 - if there is a good reason, then a new RCn_OPTION enum for ArmDisarm and rename existing to ArmDisarmAirMode would work

**Comment by CJ-Davies** [comment]
> * why would you not want airmode?

I opened the original issue #17836.

I fully appreciate the utility of airmode for performing zero throttle manoeuvres, as I have had it permanently enabled on all of my Betaflight FPV quads for years. However it should not be _forced_ upon copters that never envisage performing such manoeuvres.

The simple fact that the Arducopter implementation of airmode causes the motors to always idle at the higher value of MOT_SPIN_MIN even while on the ground is enough for me to not want it if I'm not actually using it. Arming a copter only to have the motors jump up to several % more than you expect is not a pleasant experience, nor is having the motors jump up several % if you arm in (for example) loiter but then decide to drop into stabilize to perform a takeoff.

If somebody has no expectation of using a feature, they shouldn't be forced to have it enabled when it has the potential to cause issues or unexpected behaviour.

> this seems like a dangerous for existing users who may find their motors stop when they descend fast

Surely we can introduce the ability to explicitly disable airmode on switch arming, but keep the default behaviour as enabled on switch arming?

**Comment by MichelleRos** [comment]
> Arming a copter only to have the motors jump up to several % more than you expect is not a pleasant experience

This has actually taken me by surprise a few times as well, and I agree that it's quite unnerving... When the throttle suddenly spikes higher than I thought it would, I have no idea whether it's about to go full throttle or something. A few times I've nearly disarmed before I realised it's only airmode.

In part this has more to do with airmode causing the higher "idle" speed than with airmode's function... AFAIK in Betaflight they only turn on airmode when the throttle first goes above a few percent, which would also help.

**Comment by CJ-Davies** [comment]
Airmode is definitely active in Betaflight at 0% throttle, that's its purpose. My understanding is that Betaflight simply doesn't have the equivalent of MOT_SPIN_MIN, it only has the equivalent of MOT_SPIN_IDLE.

I have never fully understood the reason for MOT_SPIN_MIN in Copter. It is defined in the doc as _'Point at which the thrust starts'_ & I have seen various people explain that it is used to ensure that the motors never slow down while in flight so much that the copter can no longer maintain attitude. This makes no sense to me though, as surely the attitude correction routine simply sets the speeds of the motors to whatever they actually need to be, as & when required?

Does MOT_SPIN_MIN actually just function as a binary control that enables/disables attitude control & artificially raises the idle throttle for no reason? Or does it perform some other function that I'm not aware of?

**Comment by MichelleRos** [comment]
> Airmode is definitely active in Betaflight at 0% throttle, that's its purpose

What I meant is it activates when you _first_ go above a certain throttle %, but then it remains on until you disarm.

**Comment by CJ-Davies** [comment]
Ah I see what you mean now. Yes, this explains the common anomaly when bench testing without props, where if you arm & then touch the throttle, it will gradually rise of its own accord.

**Comment by andyp1per** [comment]
> Does MOT_SPIN_MIN actually just function as a binary control that enables/disables attitude control & artificially raises the idle throttle for no reason? Or does it perform some other function that I'm not aware of?

I believe its to prevent motor stalls and the latency involved in regaining control after a stall. On smaller motors this is unlikely but on bigger motors more of a problem.

**Comment by CJ-Davies** [comment]
That makes sense.

But now I'm wondering what the justification is for airmode increasing MOT_SPIN_ARM to MOT_SPIN_MIN in the first place? Presumably on a copter where airmode _isn't_ being used, MOT_SPIN_MIN will be enforced once the autopilot has detected a takeoff & will remain enforced until it detects a landing. What does enabling airmode change about this scenario that would require MOT_SPIN_MIN to be enforced at all times (eg even while landed)?

File: libraries/RC_Channel/RC_Channel.h [hunk 1]
```diff
@@ -213,6 +213,7 @@ class RC_Channel {
         // options 150-199 continue user rc switch options
         CRUISE =             150,  ///CRUISE mode
         TURTLE =             151,  // Turtle mode - flip over after crash
+        ARMDISARM_WITH_AIRMODE = 152, // arm or disarm vehicle enableing airmode
```
Comment (peterbarker) [inline libraries/RC_Channel/RC_Channel.h:216]: To not change existing behaviour we must rename the old number to `ARMDISARM_WITH_AIRMODE` and create a new `ARMDISARM` here

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/RC_Channel/RC_Channel.h (hunk 1 above)
Comment (IamPete1) [reply libraries/RC_Channel/RC_Channel.h:216]: I think changing the existing behavior is the correct thing to do. 

Getting airmode when you don't want it can cause unexpected takeoffs and make disarming hard, generally quite nasty things that are not easy to fix, you cant just drop throttle to 0, would would have to have a go switch your disarm switch.

Not having airmode when you expected it causes you to go "oops" and instinctively raise throttle at which point stabilization is active again. 

We will have to add it to release notes so that those who wanted airmode on arming can get it back again.

Comment (peterbarker) [reply libraries/RC_Channel/RC_Channel.h:216]: > I think changing the existing behavior is the correct thing to do.

Not convinced :-)

> Getting airmode when you don't want it can cause unexpected takeoffs and make disarming hard, generally quite nasty things that are not easy to fix, you cant just drop throttle to 0, would would have to have a go switch your disarm switch.

Unexpectedly not having airmode because you upgraded your firmware is likely to cause unexpected crashes :-)

> Not having airmode when you expected it causes you to go "oops" and instinctively raise throttle at which point stabilization is active again.
> 
> We will have to add it to release notes so that those who wanted airmode on arming can get it back again.

Still not getting this.  Users that already have airmode on are used to the behaviour they have.  They may choose to switch to the "no airmode" option - if they read the release notes and realise it exists.  New users would have the option of choosing one behaviour or the other according to their preferences/guesses.

I just don't see any advantages in not taking the conservative upgrade path.

Comment (IamPete1) [reply libraries/RC_Channel/RC_Channel.h:216]: > Unexpectedly not having airmode because you upgraded your firmware is likely to cause unexpected crashes :-)

I disagree, this only changes acro and stabilize, in which case your actively flying the vehicle and spot it is not doing what you expect and instinctively raise the throttle avoiding the crash. 

> Still not getting this. Users that already have airmode on are used to the behaviour they have. They may choose to switch to the "no airmode" option - if they read the release notes and realise it exists. New users would have the option of choosing one behaviour or the other according to their preferences/guesses.

I'm thinking of the users that have no idea what airmode is and just currently have it by accident, for them they don't know that they have airmode so cant take that decision to turn it off. 

TLDR: I think having airmode when you didn't want it is worse than not having airmode when you did.

Comment (CJ-Davies) [reply libraries/RC_Channel/RC_Channel.h:216]: > TLDR: I think having airmode when you didn't want it is worse than not having airmode when you did.

100% agree with this, as somebody who has been using airmode for many years.

Airmode can cause unexpected behaviour, which can result in damage or injury. It should be a conscious, explicit decision to activate it, it shouldn't be bundled as a side-effect of activating something else. I for one didn't even realise until too late that switch arming would also enable airmode, which still seems like a poor decision that should now be addressed with this commit.

**Comment by IamPete1** [comment]
https://github.com/ArduPilot/ardupilot/pull/18590

**Comment by IamPete1** [comment]
I do also think that we should change so that all airmodes are the same, currently airmode via dedicated switch  is a little different to airmode via interlock or arming switch. Grep copter for `air_mode`, there are lots of uses of that beyond where `armed_with_airmode_switch` is checked.

**Review by rmackay9 — APPROVED** [review]

**Comment by IamPete1** [comment]
Fixed conflicting auxmode number.

File: ArduCopter/RC_Channel.cpp [hunk 2]
```diff
@@ -583,6 +574,13 @@ bool RC_Channel_Copter::do_aux_function(const aux_func_t ch_option, const AuxSwi
             }
             break;
 
+        case AUX_FUNC::ARMDISARM_WITH_AIRMODE:
```
Comment (peterbarker) [inline ArduCopter/RC_Channel.cpp:577]: Why not just rename the virtual method instead?

**Review by peterbarker — COMMENTED** [review]
Where are the same patches in Plane?

I would expect to see the enumeration value (https://github.com/ardupilot/ardupilot/blob/master/ArduPlane/AP_Arming.cpp#L215) change to preserve existing behaviour.

File: ArduCopter/RC_Channel.cpp (hunk 2 above)
Comment (IamPete1) [reply ArduCopter/RC_Channel.cpp:577]: Both `ARMDISARM_WITH_AIRMODE` and `ARMDISARM` call the same function for all the other vehicles. I could have renamed the method and just have it local to copter I guess and call here. 

This highlights a issue that we did not consider when deciding not to change the behaviour. In keeping it the same on copter all other vehicles are now using `ARMDISARM_WITH_AIRMODE` despite that airmode on arming functionality only existing on copter. 

I guess plan C could be to do back to the change in behaviour but do the param conversion on copter.

**Comment by IamPete1** [comment]
> Where are the same patches in Plane?
> 
> I would expect to see the enumeration value (https://github.com/ardupilot/ardupilot/blob/master/ArduPlane/AP_Arming.cpp#L215) change to preserve existing behaviour.

There is no change in behaviour on plane, it has a param to enable so both the old switch arming and new remain the same act the same. I have not bothered to add the new option to the param description for plane, people might as well carry on using the same one. Of course the new one will work exactly the same.

File: ArduCopter/RC_Channel.cpp [hunk 3]
```diff
@@ -583,6 +574,13 @@ bool RC_Channel_Copter::do_aux_function(const aux_func_t ch_option, const AuxSwi
...
+            RC_Channel::do_aux_function_armdisarm(ch_flag);
+            if (copter.arming.is_armed()) {
+                copter.ap.armed_with_airmode_switch = true;
```
Comment (andyp1per) [inline ArduCopter/RC_Channel.cpp:580]: Shouldn't this have } else { .. = false ?

**Comment by peterbarker** [comment]
On Sun, 12 Sep 2021, Peter Hall wrote:

>       Where are the same patches in Plane?
>
>       I would expect to see the enumeration value (https://github.com/ardupilot/ardupilot/blob/master/ArduPlane/AP_Arming.cpp#L215) change to preserve existing behaviour.
> 
> There is no change in behaviour on plane, it has a param to enable so both the old switch arming and new remain the same act the same. I have not bothered to add the new option to the param description for plane, people might as well carry on using the same one. Of course the new one will work exactly the same.

The airmode-on-switch in Plane has been implemented somewhere it doesn't 
have access to the information required to make a decision to leave 
airmode off.

We could fix that, or, as you say, just have the two RC options mean the 
same on Plane.  We will lost the opportunity to give the user the option 
of one or the other.

File: ArduCopter/RC_Channel.cpp (hunk 3 above)
Comment (IamPete1) [reply ArduCopter/RC_Channel.cpp:580]: its always set false on arming, see  `ArduCopter/AP_Arming.cpp` above

**Comment by IamPete1** [comment]
@peterbarker On plane you have to go and enable it with a param, we could change that so airmode with a switch works but then we would have to do a param conversion somewhere, either on plane or copter.

**Comment by peterbarker** [comment]
> @peterbarker On plane you have to go and enable it with a param, we could change that so airmode with a switch works but then we would have to do a param conversion somewhere, either on plane or copter.

Having only the old one work might allow us to dig our way out of that hole later.  Reserving that as a "fail" should just be a block in Plane's RC_Channel library which handles the aux function before the library does - and just return MAV_RESULT_FAILED or whatever.

**Comment by IamPete1** [comment]
Thinking about it there is no way we can do param conversion, so were stuck like this I think.

**Comment by IamPete1** [comment]
@peterbarker I had forgotten about button... I don't think we need the param conversion there, its not gone out in stable..... yet.

**Review by tridge — APPROVED** [review]

**Review by peterbarker — APPROVED** [review]
